WIP: Rich structured error handling (#460)

bendk opened the pull request:
This is definitely not ready to merge, but I think it's close enough to support a discussion on the general approach.  The Rust, Python, Ruby, and Kotlin code are working well enough for the coverall tests to pass.  The next step would be to implement swift.  I'm also hoping to simplify the template code far creating function calls, like @rfk suggests in the issue.

bendk commented on uniffi/src/ffi/rustcalls.rs:
```diff
@@ -0,0 +1,218 @@
+///
+/// - The consumer code allocates a `RustCallStatus`
+///    - The error_buf field can either be uninitialized or initialized to zero-capacity and a null
+///      data buffer.  The data here is going to be overwritten so any allocations will leak memory.
```
So basically the way we're using `RustBuffer` is:
  - The consumer passes in a zeroed out rust buffer struct
  - In the case of an Error, we overwrite that data with a `RustBuffer` that contains the error.

Is this usage supported or should it be?  If so I think we should update the `RustBuffer` documentation to mention that.  If not, is there a better way to achieve this?

bendk replied:
That makes a lot of sense.  For some reason I was thinking that `RustBuffer`s were intended to be immutable, but after re-reading the docs I realize that's not the case.  Is it generally true that passing in a `RustBuffer` to a function means transferring ownership and therefore the calling code shouldn't try to access it once they pass it in?  If so, maybe I'll add a bullet point to the `RustBuffer` doc comment about that.

bendk commented on uniffi/src/ffi/rustcalls.rs:
```diff
@@ -0,0 +1,218 @@
+fn make_call<F, R>(out_status: &mut RustCallStatus, callback: F) -> R::Value
+where
+    F: panic::UnwindSafe + FnOnce() -> Result<R, RustBuffer>,
+    R: IntoFfi,
```
I would really love to get rid of the dependency on `ffi_support::InfoFfi` here.  The only thing we're using it for is to define a value that we can return for errors/panics, which the consumer code will just throw away.  I can think of a couple ways, but I'm not sure if they make sense and are in scope for this project:

  - Return a `MaybeUninit<>` value.  This feels like the cleanest solution, but then it would require quite a few changes to the scaffolding templates.
  - Add a `uniffi` trait that defines a error return value for each `ViaFFI::FfiType`.  One issue with this is it would be a breaking change for any consumers that defined their own `ViaFfi` impls with other `FfiType` types.

rfk replied:
Hrm, interesting. I think you're right the `MaybeUninit` feels like the right thing, but I don't have a good sense of how disruptive that would be.

I suppose an alternative would be to do what Rust does under the hood, and actually return a `union` that can contain the error buffer *or* the success value. But that feels likely to be clunky for common cases where we're returning primitive values.

bendk replied:
I thought about a union and it could be very nice.  I think the fact that it's going to be sized larger than most primitives should be fine since we get to avoid creating the `RustCallStatus` out param.  The issue I couldn't figure it out was how to deal with all the different return types:
  - You could make a separate union for each return type `union U8Return { value: u8, error: RustBuffer }`, but this is going to be a lot of unions both on the scaffolding and bindings side.
  - You could make a single union that covers all return types, but then if 1 return type is very large then the entire union is going to be inefficient.

One question I had is if we could make any assumptions on which `FfiType` instances can be returned.  I think uniffi only uses integers, `void *`, and `RustBuffer`.  Can we assume that the consumer code isn't going to use those too, or could they return some exotic type?

rfk replied:
> One question I had is if we could make any assumptions on which FfiType instances can be returned. I think uniffi
> only uses integers, void *, and RustBuffer. Can we assume that the consumer code is going to use those too, or could
> they return some exotic type?

This is probably true in practice right now, but architecturally it feels like we'll make a better longer-term design if we do not assume this.

I also think it would be totally fine to punt this to followup work and continue to use `IntoFfi` for the first round.

bendk commented on uniffi_bindgen/src/bindings/ruby/templates/ErrorTemplate.rb:
```diff
@@ -1,50 +1,89 @@
+{%- let canonical_type_name = typ.canonical_name()|class_name_rb %}
+  {{ e.name()|class_name_rb }} => :read{{ canonical_type_name }},
+{%- endfor %}
+}
```
Would it be a good idea to make `error_module_to_reader_method` private somehow or just flagging it internal like in Python?  I don't know nearly enough about Ruby here.

bendk replied:
My second attempt is to make the variable name upper case (`ERROR_MODULE_TO_READER_METHOD`).  That makes the code work, although I'm still not sure if it's idomatic Ruby.

saks replied:
I had a quick look. Having this mapping defined as a constant `ERROR_MODULE_TO_READER_METHOD` is definitely a good idea and quite idiomatic for ruby as for me. There are ways to mark constants/classes/modules as private but language still allows these to be accessed with simple meta-programming techniques.

I'll spend more time on this later today.
Thanks!

saks replied:
I marked few constants as private @bendk, please have a look https://github.com/bendk/uniffi-rs/pull/1

bendk replied:
@saks that's wonderful, thanks so much!  I just merged them into the branch I'm working on, they'll be in the final product.

bendk commented on uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt:
```diff
@@ -1,149 +1,95 @@
+// Helpers for calling Rust
 // In practice we usually need to be synchronized to call this safely, so it doesn't
 // synchronize itself
-private inline fun <U, E: RustErrorReference> nullableRustCall(callback: (E) -> U?, err: E): U? {
```
As far as I could tell, we never called `nullableRustCall` so I removed support for it.  Tell me if that's a bad idea.

bendk commented:
[coverall-generated-code.zip](https://github.com/mozilla/uniffi-rs/files/6811735/coverall-generated-code.zip)

Here's a copy of the generated code for the coverall fixture.  I found it much easier to reason about the error handling templates when I could see the output.

rfk reviewed (commented):
Thanks @bendk, some initial comments below.

bendk commented on fixtures/coverall/tests/bindings/test_coverall.py:
```diff
@@ -75,7 +75,7 @@ def test_constructors(self):
             Coveralls.fallible_new("", True)
 
-        with self.assertRaisesRegex(InternalError, "expected panic: woe is me"):
+        with self.assertRaisesRegex(InternalError, "Rust panic"):
```
Oops, meant to flag this but I forgot.  I was worried about what happens if the code to extract the string itself panics.  For example, it seems like the current code could panic: https://github.com/mozilla/ffi-support/blob/31ffbbd6eec90038d825dab145c5382ca248b2f4/src/error.rs#L257.  I think calling `to_string()` on a `str` shouldn't panic.  But couldn't calling `clone()` on a `String` panic when trying to allocate the new string?  There's also the issue of converting the `String` to a `RustBuffer`, which I think should be panic-free, but I'm not sure about that.

I was hoping that the panic hook could log enough info to debug things, maybe as a Sentry breadcrumb.  I'm not sure if this will work though.  (BTW, what happens if the panic hook panics?  I couldn't find any documentation on this).

If that's not possible, then I think we should put the code to extract the panic string in a second `catch_unwind` block.

rfk replied:
Yes, you're right, I believe at least the allocation underlying `String::clone` could fail in this case and trigger a panic.

> I was hoping that the panic hook could log enough info to debug things, maybe as a Sentry breadcrumb.

We haven't had good luck getting rust-level debugging output into Sentry, although it's probably time we tried again. 

> I think we should put the code to extract the panic string in a second catch_unwind block.

Making a best-effort attempt to get the string out along these lines seems valuable to me 👍🏻

bendk commented on uniffi_bindgen/src/bindings/python/templates/RustBufferStream.py:
```diff
@@ -163,6 +163,34 @@ def read{{ canonical_type_name }}(self):
+    def read{{ canonical_type_name }}(self):
+        variant = self._unpack_from(4, ">i")
+        try:
+            read_variant_method = getattr(self, 'read{{canonical_type_name}}{}'.format(variant))
```
I forgot about this one, but since `variant` is an integer here,  maybe it's okay.  I don't think `MyEnumeration::1` is a valid name.

bendk commented:
@rfk Thanks for all the review!  I just pushed some changes based on your comments:

 - Changed `STATUS_SUCCESS` to be 0
 - Specified that the caller should initialize `RustCallStatus` to STATUS_SUCCESS and an empty rust buffer
 - Fixed a bunch of spelling issues
 - Updated the changelog
 - Deleted `ErrorTemplate.rs`
 
TODO:
  - Implement swift
  - Figure out how to handle panics messages
  - Figure out the error values

saks commented on fixtures/coverall/tests/bindings/test_coverall.rb:
```diff
@@ -80,10 +80,6 @@ def test_constructors
 
-    assert_raise_message /expected panic: woe is me/ do
-      Coverall::Coveralls.panicing_new('expected panic: woe is me')
-    end
```
FYI: this test passes for me on this branch if I change it to:
```ruby
    assert_raise_message 'Rust panic' do
      Coverall::Coveralls.panicing_new('Rust panic')
    end
```

bendk replied:
The original reasoning was that I couldn't match the message anymore and I thought there was another test case for simply checking if the exception was raised.  I'm going to try to re-implement the panic message functionality then add this test back.

bendk commented:
Thanks for all the help with this one.  I think the branch is getting close to ready.  I'm going to close this one and open a new PR for it.
